tests/metrics/functional/test_classification.py

```diff
@@ -337,9 +337,55 @@ def test_iou(half_ones, reduction, remove_bg, expected):
     target = (torch.arange(120) % 3).view(-1, 1)
     if half_ones:
         pred[:60] = 1
-    iou_val = iou(pred, target, remove_bg=remove_bg, reduction=reduction)
+    iou_val = iou(
+        pred=pred,
+        target=target,
+        remove_bg=remove_bg,
+        reduction=reduction,
+    )
     assert torch.allclose(iou_val, expected, atol=1e-9)
 
 
+@pytest.mark.parametrize(['pred', 'target', 'not_present_score', 'num_classes', 'remove_bg', 'expected'], [
+    # Note that -1 is used as the not_present_score in almost all tests here to distinguish it from the range of valid
+    # scores the function can return ([0., 1.] range, inclusive).
+    # 2 classes, class 0 is correct everywhere, class 1 is not present.
+    pytest.param([0], [0], -1., 2, False, [1., -1.]),
+    pytest.param([0, 0], [0, 0], -1., 2, False, [1., -1.]),
+    # not_present_score not applied if only class 0 is present and it's the only class.
+    pytest.param([0], [0], -1., 1, False, [1.]),
+    # 2 classes, class 1 is correct everywhere, class 0 is not present.
+    pytest.param([1], [1], -1., 2, False, [-1., 1.]),
+    pytest.param([1, 1], [1, 1], -1., 2, False, [-1., 1.]),
+    # When background removed, class 0 does not get a score (not even the not_present_score).
+    pytest.param([1], [1], -1., 2, True, [1.0]),
+    # 3 classes. Only 0 and 2 are present, and are perfectly predicted. 1 should get not_present_score.
+    pytest.param([0, 2], [0, 2], -1., 3, False, [1., -1., 1.]),
+    pytest.param([2, 0], [2, 0], -1., 3, False, [1., -1., 1.]),
+    # 3 classes. Only 0 and 1 are present, and are perfectly predicted. 2 should get not_present_score.
+    pytest.param([0, 1], [0, 1], -1., 3, False, [1., 1., -1.]),
+    pytest.param([1, 0], [1, 0], -1., 3, False, [1., 1., -1.]),
+    # 3 classes, class 0 is 0.5 IoU, class 1 is 0 IoU (in pred but not target; should not get not_present_score), class
+    # 2 is not present.
+    pytest.param([0, 1], [0, 0], -1., 3, False, [0.5, 0., -1.]),
+    # 3 classes, class 0 is 0.5 IoU, class 1 is 0 IoU (in target but not pred; should not get not_present_score), class
+    # 2 is not present.
+    pytest.param([0, 0], [0, 1], -1., 3, False, [0.5, 0., -1.]),
+    # Sanity checks with not_present_score of 1.0.
+    pytest.param([0, 2], [0, 2], 1.0, 3, False, [1., 1., 1.]),
+    pytest.param([0, 2], [0, 2], 1.0, 3, True, [1., 1.]),
+])
+def test_iou_not_present_score(pred, target, not_present_score, num_classes, remove_bg, expected):
+    iou_val = iou(
+        pred=torch.tensor(pred),
+        target=torch.tensor(target),
+        not_present_score=not_present_score,
+        num_classes=num_classes,
+        remove_bg=remove_bg,
+        reduction='none',
+    )
+    assert torch.allclose(iou_val, torch.tensor(expected).to(iou_val))
```

Comment:
These tests are fine, but could you also write a test that on random data test against sklearns `sklearn.metrics.jaccard_score` (jaccard score and iou is the same by my understanding)

Reply:
Ah, another good point. In https://github.com/PyTorchLightning/pytorch-lightning/pull/3098/commits/b0057d1c5dc8728496fafa24f143e4eb8c0a6027, I add a comparison against sklearn's jaccard score to the existing `test_against_sklearn` above.

I also added a TODO note for comparing our `not_present_score` with sklearn's jaccard score's `zero_division` parameter being introduced in https://github.com/scikit-learn/scikit-learn/pull/17866 whenever that becomes available. This would likely be a specific new test block not rolled into any of the existing sklearn tests.

Reply:
Marking this conversation as resolved, but feel free to re-open for further discussion or changes.